design/20210715.bundle-crd.md

```diff
@@ -0,0 +1,275 @@
+
+Bundle is intended to be powerful and flexible enough to meet the use case of
+most users who want some set of trust bundle data available for their services.
+This trust bundle data may or may not be known to the cluster operator before
+runtime.
+
+## Motivation
+
+Kubernetes currently has a limited story for distributing trust within a
+cluster. While API server CA bundle data is available to all pods at a
+well-known
+[file location](https://kubernetes.io/docs/tasks/run-application/access-api-from-pod/#directly-accessing-the-rest-api)
+mounted by the Kubelet at runtime, as well as the
+[ConfigMap in all namespaces `kube-root-ca.crt`](https://github.com/kubernetes/kubernetes/pull/100926/files#diff-60eb175dae1dcc1320f5c426f6d193a67e961e2bd96d574f941cdf09672090f6R44),
+this bundle data can only be modified via
+[API server CLI flags](https://kubernetes.io/docs/setup/best-practices/certificates/).
+Modifying these CLI flags to append arbitrary trust bundles is both impossible
+for managed Kubernetes users, and is a misuse of its intention. CLI flags are
+static, and can't be realistically changed in response to runtime events.
+
+cert-manager exposes a
+[`Status.CA` field on CertificateRequests](https://cert-manager.io/docs/concepts/certificaterequest/),
+as well as stores a
+[`ca.crt` field in generated Secrets resulting from Certificates](https://cert-manager.io/docs/concepts/certificate/).
+These values are however: best-effort by the Issuer and not required for a
+successful issuance; only available after a successful certificate request and
+scoped to that namespace; regarded as a controversial inclusion to the
+cert-manager API. This field only ever in practice contains one CA certificate,
+making managing multiple CA certificates or bundles difficult or impossible.
+Services often need multiple CA certificates as part of the same bundle or file.
+
+The upstream
+[Kubernetes CertificateSigningRequest resource](https://kubernetes.io/docs/reference/access-authn-authz/certificate-signing-requests/)
+does not include any CA field. This means consumers after forced,
+[and are intended to](https://kubernetes.io/docs/reference/access-authn-authz/certificate-signing-requests/#kubernetes-signers),
+find an out of band method of propagating CA data.
+
+Some users have success by baking in CA data within built container images.
+This however requires the assumption that CA data is: known before runtime, long
+lived, and all images that require the CA data are built in house.
+
+Projects such as [Istio](https://istio.io/) (as well as Kubernetes itself) has
+had success with distributing trust roots, at runtime, by using a
+[well-known ConfigMap `istio-ca-root-cert`](https://istio.io/latest/docs/tasks/security/cert-management/plugin-ca-cert/)
+that is available within every namespace in a cluster. Bundle is an extension of
+this idea- to allow users to use the same approach with a resource that is
+expressive enough to meet most needs.
+
+## Goals
```

Comment:
question: :warning: I think this is an important one

Something that I don't think is mentioned in this document is updating a trust store. It seems like a hard problem, because I think in practice an update would require a cluster operator to roll every deployment which consumed the updated trust store, but equally it's an important thing to have.

Without the ability to update a trust store, we can't:

- rotate an expiring root
- replace/remove an exposed certificate
- easily update publicly-trusted CAs lists (if supported), which are regularly changed

It feels like we need a story on this; do you have any thoughts on it?

Reply:
>  Something that I don't think is mentioned in this document is updating a trust store.

This is left up to the cluster operator and shouldn't be something that Bundle is concerned with. Bundle will ensure all targets are up to date at all times, but the cluster operator must use either: 1. use moderns apps that watch for changes to trust data files, or 2. use some custom solution like a sidecar or manual rollout.

Reply:
That's a fair position to take; it might be worth adding something to that effect into the doc

Reply:
Yes, I agree.